// react-formal/es/NestedForm.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import PropTypes from 'prop-types';
import React from 'react';
import Form from './Form';
import Field from './Field';
import { filter, prefix, unprefix } from './utils/ErrorUtils';
/**
 * A `Form` component that takes a `name` prop. Functions exactly like a normal
 * Form, except that when a `name` is present it will defer errors up to the parent `<Form>`,
 * functioning like a `<Form.Field>`.
 *
 * This is useful for encapsulating complex input groups into self-contained
 * forms without having to worry about `"very.long[1].paths[4].to.fields"` for names.
 */

var NestedForm =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(NestedForm, _React$Component);

  function NestedForm() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = NestedForm.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        name = _this$props.name,
        schema = _this$props.schema,
        errors = _this$props.errors,
        props = _objectWithoutPropertiesLoose(_this$props, ["name", "schema", "errors"]);

    return React.createElement(Field, {
      name: name,
      noResolveType: true,
      noValidate: true,
      events: "onChange"
    }, function (_ref) {
      var meta = _ref.meta,
          value = _ref.value,
          onChange = _ref.onChange;
      return React.createElement(Form, _extends({
        as: "div"
      }, props, {
        value: value,
        onChange: onChange,
        onError: function onError(errors) {
          return meta.onError(prefix(errors, name));
        },
        errors: unprefix(name ? meta.errors : errors, name),
        schema: schema || meta.schema,
        context: name ? _extends({}, meta.context, props.context) : props.context
      }));
    });
  };

  return NestedForm;
}(React.Component);

NestedForm.propTypes = {
  name: PropTypes.string.isRequired,
  schema: PropTypes.object,
  errors: PropTypes.object,
  meta: PropTypes.shape({
    errors: PropTypes.object.isRequired,
    onError: PropTypes.func.isRequired
  })
};
export default NestedForm;
