// react-formal/es/Summary.js
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import React from 'react';
import PropTypes from 'prop-types';
import elementType from 'prop-types-extra/lib/elementType';
import Message from './Message';
/**
 * Display all Form validation `errors` in a single summary list.
 *
 * ```jsx { "editable": true }
 * <Form
 *   schema={modelSchema}
 *   defaultValue={modelSchema.default()}
 * >
 *   <Form.Summary/>
 *
 *   <Form.Field name='name.first' placeholder='first'/>
 *   <Form.Field name='name.last' placeholder='surname'/>
 *   <Form.Field name='dateOfBirth' placeholder='dob'/>
 *
 *   <Form.Submit>Validate</Form.Submit>
 * </Form>
 * ```
 */

var Summary =
/*#__PURE__*/
function (_React$PureComponent) {
  _inheritsLoose(Summary, _React$PureComponent);

  function Summary() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Summary.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        formatMessage = _this$props.formatMessage,
        props = _objectWithoutPropertiesLoose(_this$props, ["formatMessage"]);

    return React.createElement(Message, props, function (errors) {
      return errors.map(formatMessage);
    });
  };

  return Summary;
}(React.PureComponent);

Summary.propTypes = {
  /**
   * An error message renderer, Should return a `ReactElement`
   * ```
   * function(
   *   message: string,
   *   idx: number,
   *   errors: array
   * ) -> ReactElement
   * ```
   */
  formatMessage: PropTypes.func.isRequired,

  /**
   * A DOM node tag name or Component class the Message should render as.
   */
  as: elementType.isRequired,

  /**
   * A css class that should be always be applied to the Summary container.
   */
  errorClass: PropTypes.string,

  /**
   * Specify a group to show errors for, if empty all form errors will be shown in the Summary.
   */
  group: PropTypes.string
};
Summary.defaultProps = {
  as: 'ul',
  formatMessage: function formatMessage(message, idx) {
    return React.createElement("li", {
      key: idx
    }, message);
  }
};
export default Summary;
